File: api/app/api/meal.py
from typing import Any
from corepy.api.result import ApiResult, IdResult, ListResult
from fastapi.routing import APIRouter

from app.config import SessionDep
from app.model import MealCreateParam, MealListParam, MealSimplePublic, MealUpdateParam, TagSimplePublic
from app.repo.meal import create_meal, delete_meal, get_meal, list_meals, update_meal
from app.service.meal import list_tags
from app.common import BizException

router = APIRouter(prefix="/meal", tags=["meal"])

@router.post("/list", response_model=ApiResult[ListResult[MealSimplePublic]])
def _list(session: SessionDep, params: MealListParam) -> Any:
    entities = list_meals(session, params)
    res = [MealSimplePublic.create(entity) for entity in entities]
    return ApiResult.ok(ListResult.create(res))

@router.post("/get", response_model=ApiResult[MealSimplePublic])
def _get(session: SessionDep, id: str) -> Any:
    entity = get_meal(session, id)
    print(f"entity {entity}")
    if not entity:
        raise BizException(f"meal not found: {id}")
    return ApiResult.ok(MealSimplePublic.create(entity))

@router.post("/create", response_model=ApiResult[IdResult])
def _create(session: SessionDep, params: MealCreateParam) -> Any:
    res = create_meal(session, params)
    return ApiResult.ok(IdResult(id=res))

@router.post("/update", response_model=ApiResult[None])
def _update(session: SessionDep, params: MealUpdateParam) -> Any:
    if not update_meal(session, params):
        raise BizException(f"meal not found: {params.id}")
    return ApiResult.ok()

@router.post("/delete", response_model=ApiResult[None])
def _delete(session: SessionDep, id: str) -> Any:
    delete_meal(session, id)
    return ApiResult.ok()


@router.post("/tag/list", response_model=ApiResult[ListResult[TagSimplePublic]])
def _tag_list(session: SessionDep, name: str | None = None) -> Any:
    res = list_tags(session, name)
    return ApiResult.ok(ListResult.create(res))

File: api/src/api/meal.rs
use axum::{Json, Router, extract::State, routing::post};
use rexl::api::{ApiResult, PageResult};

use crate::config::ApiState;
use crate::model::{
    IdParam, IdResult, MealCreateParam, MealListParam, MealSimplePublic, MealUpdateParam,
    TagSimplePublic,
};
use crate::service::MealService;
use crate::util::AppResult;

pub fn get_routes() -> Router<ApiState> {
    Router::new().nest(
        "/meal",
        Router::new()
            .route("/list", post(list_meal))
            .route("/get", post(get_meal))
            .route("/create", post(create_meal))
            .route("/update", post(update_meal))
            .route("/delete", post(delete_meal))
            .route("/tag/list", post(list_tags)),
    )
}

async fn list_meal(
    State(state): State<ApiState>, Json(params): Json<MealListParam>,
) -> AppResult<Json<ApiResult<PageResult<MealSimplePublic>>>> {
    let items = MealService::list_meals(&state, &params).await?;
    let total = items.len() as u64;
    Ok(Json(ApiResult::page(total, items)))
}

async fn get_meal(
    State(state): State<ApiState>, Json(params): Json<IdParam>,
) -> AppResult<Json<ApiResult<MealSimplePublic>>> {
    let entity = MealService::get_meal(&state, &params.id).await?;
    Ok(Json(ApiResult::ok(entity)))
}

async fn create_meal(
    State(state): State<ApiState>, Json(params): Json<MealCreateParam>,
) -> AppResult<Json<ApiResult<IdResult>>> {
    let id = MealService::create_meal(&state, params).await?;
    Ok(Json(ApiResult::ok(IdResult { id })))
}

async fn update_meal(
    State(state): State<ApiState>, Json(params): Json<MealUpdateParam>,
) -> AppResult<Json<ApiResult<()>>> {
    MealService::update_meal(&state, params).await?;
    Ok(Json(ApiResult::<()>::ok(None)))
}

async fn delete_meal(
    State(state): State<ApiState>, Json(params): Json<IdParam>,
) -> AppResult<Json<ApiResult<()>>> {
    MealService::delete_meal(&state, &params.id).await?;
    Ok(Json(ApiResult::<()>::ok(None)))
}

async fn list_tags(
    State(state): State<ApiState>,
) -> AppResult<Json<ApiResult<PageResult<TagSimplePublic>>>> {
    let items = MealService::list_tags(&state).await?;
    let total = items.len() as u64;
    Ok(Json(ApiResult::page(total, items)))
}
